### DISCUSSION TITLE
Embed spotless

### DISCUSSION THREAD

**Discussion by snicoll** [diffplug/spotless#1705] [Q&A] [answered]
Hey! I am working on a service that generates source code and we'd like a bit more control over the formatting. We do support Java, Kotlin, and Groovy. 

Our needs are pretty "basic" where our code would chose a formatter for each with basic settings. I am wondering if someone else already embedded spotless in a Java app to help me get started. Wondering also about the dependencies download based on the tools and if it can be done at build-time since we don't need that flexibility.

Thanks!

**Comment by nedtwigg** [comment, marked as answer]
> if someone else already embedded spotless in a Java app

We maintain the `lib` API with strict semver to make exactly this usecase easy.

- [Cleanthat](https://github.com/solven-eu/cleanthat) has embedded Spotless

> Wondering also about the dependencies download based on the tools and if it can be done at build-time

Yes, that can be done, you can implement `Provisioner` with a local classpath search. If you implement that, it would be a nice thing to contribute upstream to `lib`. That would take maintenance off of your plate, and probably other people would like to use it too.

- https://github.com/diffplug/spotless/issues/527#issuecomment-588397021
